### DISCUSSION TITLE
[graphql] Allow filtering children of node too

### DISCUSSION THREAD

**Discussion by mathieucaroff** [gatsbyjs/gatsby#14060] [Ideas / Feature Requests]
### Summary

Currently, in Gatsby's GraphQL, it is possible to filter children nodes of top-level "all-instances-of-a-type" fileds like "allFile", "allSite", "allPlugin", etc.

For instance, this works:

```graphql
query {
  allFile (filter: { name: {eq: "defaultIcon") {
    childImageSharp {
      fluid {
        ...GatsbyImageSharpFluid
        src
      }
    }
  }
}
```

What is not possible to do, that should be, is filtering children nodes of other nodes.

### Basic example

The below GraphQl request uses `filter` on `childrenFile`. This is not currently possible, in `gatsby@2.4.6`.

```graphql
query {
  allGithubRepo {
    nodes {
      repo
      childrenFile (filter: { name: {eq: "defaultIcon") {
        childImageSharp {
          fluid {
            ...GatsbyImageSharpFluid
            src
          }
        }
      }
  }
}
```

[According to @pieh](https://github.com/gatsbyjs/gatsby/issues/8764#issuecomment-427049426), this feature would have to be implemented inside gatsby core.

### Motivation

(M1)
It is a bit confusing that filtering is available on on top-level "all-instances of type" fields, but not on node children. This lead to issue #8764 being raised.

(M2)
On the project I'm working on, I use `createRemoteFileNode` from `gatsby-source-filesystem` to obtain images from github repositories, to use them in pages. The images can the default icon, some other icon, or a screenshot. All are children of the repository, but they'll be used in different component. On of the ideas behind GraphQL is that the response to a query should contain no more than the needed data. This is not possible without filtering

(M3) https://github.com/gatsbyjs/gatsby/issues/8764#issuecomment-426909034 (@RIP21)
> @pieh What about to add support for such filtering? Imagine you have 300 different keys under, and all of that will be fetched and added as a JSON making build of 10k pages weights around 10GB.
> Or maybe any idea how to add such thing to it ourself, like a fork.

**Comment by freiksenet** [comment]
Hi! That's a good idea. If you want to make a PR, here are some pointers.

Children/child fields are added in [this function](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/schema/schema.js#L517)

Filter parameters are created in [this function)(https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/schema/schema.js#L476).

Filter types are cached, so you can safely call `getFilterTC` again in the `addConvenienceChildrenFields` function too. You can use that then to add filter parameter to it.

**Comment by emckay** [comment]
+1 for this before it goes stale.

I can work around this for now by doing the children filtering when I create the page in `gatsby-node` and passing the filtered results to the page via context, but it feels like a bit of a hack. It'd be nice to be able to do the filtering in graphql in the page query instead.

**Comment by antonioperic** [comment]
+1 for this

**Comment by loganwishartcraig** [comment]
Another +1!

**Comment by thekevinbrown** [comment]
We could really use this too.

**Comment by LekoArts** [comment]
Please refrain from adding "+1" comments, GitHub has the reaction feature for that. Thanks!

If you want to contribute to this, please read our documentation here: https://www.gatsbyjs.org/contributing/how-to-contribute/

**Comment by martin-braun** [comment]
Please add this together with limit and skip on children. Also I think the sort should be available on children, separately, so I can configure sort for every level, instead of only for one field in all levels.

Is there any update regarding this?

**Comment by dtcbyron** [comment]
Just wondering if there is any update on this? I am building a custom source plugin that requires this functionality.
